apps/gdalalg_raster_fillnodata.cpp

```diff
@@ -0,0 +1,180 @@
+
+#include "gdalalg_raster_fillnodata.h"
+
+#include "gdal_priv.h"
+#include "gdal_utils.h"
+#include "gdal_alg.h"
+
+#include <algorithm>
+
+//! @cond Doxygen_Suppress
+
+#ifndef _
+#define _(x) (x)
+#endif
+
+/************************************************************************/
+/*   GDALRasterFillNodataAlgorithm::GDALRasterFillNodataAlgorithm()     */
+/************************************************************************/
+
+GDALRasterFillNodataAlgorithm::GDALRasterFillNodataAlgorithm() noexcept
+    : GDALAlgorithm(NAME, DESCRIPTION, HELP_URL)
+{
+
+    AddProgressArg();
+
+    AddInputFormatsArg(&m_inputFormats)
+        .AddMetadataItem(GAAMDI_REQUIRED_CAPABILITIES, {GDAL_DCAP_RASTER});
+    AddInputDatasetArg(&m_inputDataset, GDAL_OF_RASTER);
+    AddOutputDatasetArg(&m_outputDataset, GDAL_OF_RASTER);
+    AddOutputFormatArg(&m_format, /* bStreamAllowed = */ false,
+                       /* bGDALGAllowed = */ false)
+        .AddMetadataItem(GAAMDI_REQUIRED_CAPABILITIES,
+                         {GDAL_DCAP_CREATE, GDAL_DCAP_RASTER});
+
+    AddCreationOptionsArg(&m_creationOptions);
+    AddOverwriteArg(&m_overwrite);
+
+    AddBandArg(&m_band).SetDefault(m_band);
+
+    AddArg("max-distance", 'd',
+           _("The maximum distance (in pixels) that the algorithm will search "
+             "out for values to interpolate."),
+           &m_maxDistance)
+        .SetDefault(m_maxDistance)
+        .SetMetaVar("MAX_DISTANCE");
+
+    AddArg("smoothing-iterations", 's',
+           _("The number of 3x3 average filter smoothing iterations to run "
+             "after the interpolation to dampen artifacts. The default is zero "
+             "smoothing iterations."),
+           &m_smoothingIterations)
+        .SetDefault(m_smoothingIterations)
+        .SetMetaVar("SMOOTHING_ITERATIONS");
+
+    auto &mask{AddArg("mask", 0,
+                      _("Use the first band of the specified file as a "
+                        "validity mask (zero is invalid, non-zero is valid)."),
+                      &m_maskDataset)};
+
+    SetAutoCompleteFunctionForFilename(mask, GDAL_OF_RASTER);
+
+    mask.AddValidationAction(
+        [this, &mask]()
+        {
+            // Check if the mask dataset is a valid raster dataset descriptor
+            // Try to open the dataset as a raster
+            std::unique_ptr<GDALDataset> poDS(
+                GDALDataset::Open(mask.Get<std::string>().c_str(),
+                                  GDAL_OF_RASTER, nullptr, nullptr, nullptr));
+            return static_cast<bool>(poDS);
+        });
+
+    AddArg("strategy", 0,
+           _("By default, pixels are interpolated using an inverse distance "
+             "weighting (inv_dist). It is also possible to choose a nearest "
+             "neighbour (nearest) strategy."),
+           &m_strategy)
+        .SetDefault(m_strategy)
+        .SetChoices("inv_dist", "nearest");
```

Comment:
maybe normalize to "invdist" to be consistent with "gdal vector grid invdist" ? (and adapt in Run() to transform to inv_dist)